src/squlearn/qnn/qnn.py

```diff
@@ -938,55 +938,67 @@ def flatten(container):
             else:
                 raise ValueError("No execution is set!")
 
-            set_empty = False
-            if val.shape[0] == 0:
-                set_empty = True
-
-            # Swapp results into the following order:
-            # 1. different observables (op_list)
+            is_val_empty = False
+            if val.size == 0:
+                is_val_empty = True
+
+            # val has the following order:
+            # 0 -> x_inp and param_inp combined (has to be separated later)
+            # 1 -> param_op_inp (often just shape 1)
+            # 2:2+num_nested -> circuit derivatives of parameter shift (if needed)
+            #                   (num_nested = order of derivatives)
+            # 2+num_nested ->  different observables (op_list - todos for same circuit)
+            # 3+num_nested:last-1 -> observable derivatives (if needed)
+            # last : multi_output
+
+            # Swapp results into the following final order:
+            # 1. different observables (op_list - todo for same circuit)
             # 2. different input data/ encoding circuit parameters (x_inp,params) -> separated later
             # 3. different operator parameters (param_op_inp)
             # 4. different output values (multi_output)
             # 5. If there, lists of the operators (e.g. operator derivatives)
             # 6. if there, lists of the circuits (e.g. array for gradient)
+            # swapping is done in two stages, first the op_list are swapped to the first position
+            # afterward the order of each element in the op_list is adjusted
 
-            if set_empty is False:
+            # FIRST SWAP: different observables to the first place of the array
+            # swap i=2+num_nested to position 0, keep the rest in order
+            if is_val_empty is False:
                 ilist = list(range(len(val.shape)))
-
-                #             # Op_list index       # fm dict   # op dict
-                swapp_list = [ilist[2 + num_nested]] + [ilist[0]] + [ilist[1]]
-
-                length = 3 + num_nested
-                # Add multiple output data next
-                if self.multiple_output:
-                    length += 1
-                    swapp_list = swapp_list + [ilist[-1]]
-
-                # If there are lists in the operators, add them next (e.g. dfdop)
-                if len(ilist) > length:
-                    if self.multiple_output:
-                        swapp_list = swapp_list + ilist[3 + num_nested : -1]
-                    else:
-                        swapp_list = swapp_list + ilist[3 + num_nested :]
-
-                # If there are lists in the circuits, add them here (e.g. dfdp)
+                swapp_list2 = [ilist[2 + num_nested]]
```

Comment:
Is there any specific reason why this is called `swapp_list2` and later `swapp_list3`? IMO we could just call it `swap_list` everywhere. Also `ilist` would be better called `index_list` or `indices`.